File: backend/app/schemas/payment.py
"""Payment API contracts.

**Append-only.** A payment is a financial record: once written it is never edited or removed.
A mistake is corrected by posting a reversal, not by changing history. There is therefore no
update schema in this module, and none should be added.

``public_id`` (migration 0002) is the URL identity. It is also what lets a refund name the
charge it reverses -- ``ck_payments_refund_references_charge`` requires the link, and before
0002 there was no identifier a client could use to express it.

Charge and refund are one table distinguished by ``kind``, with ``amount`` always positive:
direction lives in ``kind``, never in the sign. That is the frozen schema's choice, and the
schemas below mirror it rather than reinterpreting it.
"""

from __future__ import annotations

import datetime as dt
import decimal
import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

#: Mirrors ck_payments_method_valid.
PaymentMethodLiteral = Literal[
    "card", "cash", "bank_transfer", "online_gateway", "ota_collect", "voucher"
]
#: Mirrors ck_payments_status_valid.
PaymentStatusLiteral = Literal[
    "pending", "authorized", "captured", "failed", "refunded", "partially_refunded", "cancelled"
]
#: Mirrors ck_payments_kind_valid.
PaymentKindLiteral = Literal["charge", "refund"]

#: ck_payments_amount_positive: amount > 0 for BOTH kinds.
AmountField = Annotated[decimal.Decimal, Field(gt=0, max_digits=14, decimal_places=2)]
CurrencyField = Annotated[str, Field(min_length=3, max_length=3, pattern=r"^[A-Z]{3}$")]
#: ck_payments_card_last_four_format.
CardLastFourField = Annotated[str, Field(min_length=4, max_length=4, pattern=r"^[0-9]{4}$")]


class PaymentBase(BaseModel):
    """Fields common to both kinds of posting."""

    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    amount: AmountField
    currency: CurrencyField
    method: PaymentMethodLiteral
    status: PaymentStatusLiteral = "pending"
    paid_at: dt.datetime | None = None
    provider: Annotated[str, Field(max_length=50)] | None = None
    transaction_reference: Annotated[str, Field(max_length=100)] | None = None
    failure_reason: Annotated[str, Field(max_length=500)] | None = None
    #: The last four digits only. The schema stores nothing else about a card, and nothing
    #: else may be sent: no PAN, no CVV, no expiry, no cardholder name.
    card_last_four: CardLastFourField | None = None

    @field_validator("currency", mode="before")
    @classmethod
    def _upper(cls, value: object) -> object:
        return value.upper() if isinstance(value, str) else value

    @model_validator(mode="after")
    def _captured_requires_a_settlement_time(self) -> PaymentBase:
        """Mirror ``ck_payments_captured_has_paid_at``.

        Caught at the edge so a client gets a field-level message naming ``paid_at`` rather
        than an opaque conflict from a constraint it cannot see.
        """
        if self.status == "captured" and self.paid_at is None:
            raise ValueError("paid_at is required when status is 'captured'")
        return self


class ChargeCreate(PaymentBase):
    """Payload for posting a charge against a booking.

    ``kind`` is not accepted: the endpoint determines it. A charge must NOT carry a parent
    reference (``ck_payments_refund_references_charge`` is a biconditional), and letting a
    client set ``kind`` would only create a way to violate that.
    """


class RefundCreate(PaymentBase):
    """Payload for posting a refund against an existing payment.

    The payment being reversed is named by its ``public_id`` -- the identity added in
    migration 0002 precisely so this operation can be expressed.
    """

    refunds_public_id: uuid.UUID


class PaymentResponse(BaseModel):
    """What the API returns.

    Carries the parents' public identifiers so a client can rebuild the record's URL, and
    ``refunds_public_id`` so a refund names its charge in the same terms it was created with.
    No internal BIGINT appears: not ``payments.id``, not ``booking_id``, not ``hotel_id``,
    not ``refunded_payment_id``.
    """

    model_config = ConfigDict(from_attributes=True)

    hotel_public_id: uuid.UUID
    booking_public_id: uuid.UUID
    public_id: uuid.UUID
    kind: PaymentKindLiteral
    amount: decimal.Decimal
    currency: str
    method: PaymentMethodLiteral
    status: PaymentStatusLiteral
    paid_at: dt.datetime | None
    provider: str | None
    transaction_reference: str | None
    #: Present only on a refund; null on a charge, mirroring the biconditional CHECK.
    refunds_public_id: uuid.UUID | None
    failure_reason: str | None
    card_last_four: str | None
    created_at: dt.datetime
    updated_at: dt.datetime


__all__ = [
    "ChargeCreate",
    "PaymentKindLiteral",
    "PaymentMethodLiteral",
    "PaymentResponse",
    "PaymentStatusLiteral",
    "RefundCreate",
]

File: frontend/src/types/payment.ts
import type { DecimalString } from '@/types/booking'

/**
 * The payment contracts, transcribed from `app.schemas.payment`.
 *
 * Confirmed against the live API -- `/openapi.json` plus real requests -- rather than read
 * from source alone. Four facts that check established, each of which shaped the UI:
 *
 * 1. **`amount` is always positive, for both kinds.** Direction lives in `kind`, never in the
 *    sign; that is the frozen schema's choice and `ck_payments_amount_positive` enforces it.
 *    Nothing in this application negates a refund.
 * 2. **Every `Decimal` is a JSON string**, on the way in as well as out. The request schema
 *    accepts a string for `amount`, so a figure typed by a user reaches PostgreSQL without
 *    ever being a JavaScript number.
 * 3. **`kind` is not accepted on either request.** The endpoint determines it -- a charge
 *    must not name a parent and a refund must, and one payload able to express both would be
 *    a payload able to express an invalid combination. Sending `kind` is a 422.
 * 4. **No internal key appears anywhere**: not `payments.id`, not `booking_id`, not
 *    `hotel_id`, not `refunded_payment_id`. `public_id` is the identity throughout.
 */

/** Mirrors `ck_payments_method_valid`. */
export type PaymentMethod =
  | 'card'
  | 'cash'
  | 'bank_transfer'
  | 'online_gateway'
  | 'ota_collect'
  | 'voucher'

/**
 * Mirrors `ck_payments_status_valid`.
 *
 * Seven states, and they are the *posting's* own status -- not the booking's financial
 * position. The booking-level answer ("unpaid", "partially_paid", "paid", "overpaid") is a
 * different vocabulary, derived by reconciliation, and the two must not be confused.
 */
export type PaymentStatus =
  | 'pending'
  | 'authorized'
  | 'captured'
  | 'failed'
  | 'refunded'
  | 'partially_refunded'
  | 'cancelled'

/** Mirrors `ck_payments_kind_valid`. Charges and refunds are one table split by this. */
export type PaymentKind = 'charge' | 'refund'

/** One posting against a booking, as the API returns it. */
export interface Payment {
  readonly hotel_public_id: string
  readonly booking_public_id: string
  readonly public_id: string
  readonly kind: PaymentKind
  /** Always positive. `NUMERIC(14,2)` on the wire as a string. */
  readonly amount: DecimalString
  /** The posting's own currency. A booking may carry postings in more than one. */
  readonly currency: string
  readonly method: PaymentMethod
  readonly status: PaymentStatus
  /** An instant. Required by `ck_payments_captured_has_paid_at` when status is `captured`. */
  readonly paid_at: string | null
  readonly provider: string | null
  /**
   * The processor's own reference.
   *
   * Half of the idempotency key: a partial unique index over
   * `(provider, transaction_reference) WHERE transaction_reference IS NOT NULL` means a
   * repeated webhook cannot create a second row. Verified live -- the same pair twice is a
   * 409 even when the amount differs, and omitting it entirely lets a second posting through,
   * which is what "partial" means.
   */
  readonly transaction_reference: string | null
  /** Present only on a refund; null on a charge, mirroring the biconditional CHECK. */
  readonly refunds_public_id: string | null
  readonly failure_reason: string | null
  /**
   * The last four digits, and the schema stores nothing else about a card.
   *
   * No PAN, no CVV, no expiry, no cardholder name -- there is no field for any of them, and
   * `additionalProperties: false` makes that enforceable: sending `card_number` is a 422,
   * verified live.
   */
  readonly card_last_four: string | null
  readonly created_at: string
  readonly updated_at: string
}

/**
 * `POST .../payments` -- the `ChargeCreate` payload.
 *
 * Required: `amount`, `currency`, `method`. Everything else is optional, and `status`
 * defaults to `pending` server-side.
 */
export interface ChargeRequest {
  /** A decimal STRING. Never a number: see the module docstring. */
  readonly amount: DecimalString
  readonly currency: string
  readonly method: PaymentMethod
  readonly status?: PaymentStatus
  /** ISO instant. Required by the server when `status` is `captured`. */
  readonly paid_at?: string
  readonly provider?: string
  readonly transaction_reference?: string
  readonly card_last_four?: string
}

/**
 * `POST .../payments/refunds` -- the `RefundCreate` payload.
 *
 * Identical to a charge plus `refunds_public_id`, which names the payment being reversed.
 * The server resolves that id **through the booking**, so a refund can only ever reverse a
 * payment the caller already had access to -- a payment on another booking is a 404, which is
 * also what stops this endpoint being used to probe for one.
 */
export interface RefundRequest {
  readonly amount: DecimalString
  /** Must equal the parent's currency; the server refuses anything else with a 409. */
  readonly currency: string
  readonly method: PaymentMethod
  readonly refunds_public_id: string
  readonly status?: PaymentStatus
  readonly paid_at?: string
  readonly provider?: string
  readonly transaction_reference?: string
}
